src/runtime/pkg/containerd-shim-v2/service.go

```diff
@@ -256,15 +255,14 @@ func (s *service) StartShim(ctx context.Context, opts cdshim.StartOpts) (_ strin
 	return address, nil
 }
 
-func (s *service) forward(ctx context.Context, publisher events.Publisher) {
+func (s *service) forward(ctx context.Context, publisher cdshim.Publisher) {
 	for e := range s.events {
-		ctx, cancel := context.WithTimeout(ctx, timeOut)
 		err := publisher.Publish(ctx, getTopic(e), e)
-		cancel()
 		if err != nil {
 			shimLog.WithError(err).Error("post event")
 		}
 	}
+	publisher.Close()
```

Comment:
I think the containerd is taking the charge of [closing](https://github.com/containerd/containerd/blob/e1ad7791077916aac9c1f4981ad350f0e3fce719/runtime/v2/shim/shim.go#L203-L207
) the publisher:


```golang
	ttrpcAddress := os.Getenv(ttrpcAddressEnv)
	publisher, err := NewPublisher(ttrpcAddress)
	if err != nil {
		return err
	}
	defer publisher.Close()

	ctx := namespaces.WithNamespace(context.Background(), namespaceFlag)
	ctx = context.WithValue(ctx, OptsKey{}, Opts{BundlePath: bundlePath, Debug: debugFlag})
	ctx = log.WithLogger(ctx, log.G(ctx).WithField("runtime", id))
	ctx, cancel := context.WithCancel(ctx)
	service, err := initFunc(ctx, idFlag, publisher, cancel)
	if err != nil {
		return err
	}
```

Reply:
I agree with you 😄 
By the way, do we need ctx with timeout in this function? It looks like containerd also cover it.
```go
func (l *RemoteEventsPublisher) forwardRequest(ctx context.Context, req *v1.ForwardRequest) error {
	service, err := l.client.EventsService()
	if err == nil {
		fCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
		_, err = service.Forward(fCtx, req)
		cancel()
		if err == nil {
			return nil
		}
	}
...
	// try again with a fresh context, otherwise we may get a context timeout unexpectedly.
	fCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
	_, err = service.Forward(fCtx, req)
	cancel()
	if err != nil {
		return err
	}

	return nil
}
```

Reply:
@Bevisy without clear evidence, we can keep it.